### PR TITLE
fix: Set value if not None

### PR DESCRIPTION
**Pull request by restanrm**
Check fail if it's equal to 'False' or '0' and it shouldn't. 
Multiple fields where impacted by this error.

### CODE REVIEW

File: pymisp/api.py [hunk 1]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
 
     def _set_user_parameters(self, **kwargs):
         user = {}
-        if kwargs.get('email'):
+        if kwargs.get('email') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: The email should not be empty string, 0 or False.

File: pymisp/api.py [hunk 2]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['email'] = kwargs.get('email')
-        if kwargs.get('org_id'):
+        if kwargs.get('org_id') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: same as email (should be an integer > 0)

File: pymisp/api.py [hunk 3]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['org_id'] = kwargs.get('org_id')
-        if kwargs.get('role_id'):
+        if kwargs.get('role_id') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: same as org_id

File: pymisp/api.py [hunk 4]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['role_id'] = kwargs.get('role_id')
-        if kwargs.get('password'):
+        if kwargs.get('password') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: If set, I don't see a reason to have the password empty string, 0 or False.

File: pymisp/api.py [hunk 5]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['password'] = kwargs.get('password')
-        if kwargs.get('external_auth_required'):
+        if kwargs.get('external_auth_required') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: ack

File: pymisp/api.py [hunk 6]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['external_auth_required'] = kwargs.get('external_auth_required')
-        if kwargs.get('external_auth_key'):
+        if kwargs.get('external_auth_key') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: not sure what this is, but can it be empty string?

File: pymisp/api.py [hunk 7]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['external_auth_key'] = kwargs.get('external_auth_key')
-        if kwargs.get('enable_password'):
+        if kwargs.get('enable_password') is not None:
             user['enable_password'] = kwargs.get('enable_password')
-        if kwargs.get('nids_sid'):
+        if kwargs.get('nids_sid') is not None:
             user['nids_sid'] = kwargs.get('nids_sid')
-        if kwargs.get('server_id'):
+        if kwargs.get('server_id') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: int > 0

File: pymisp/api.py [hunk 8]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['server_id'] = kwargs.get('server_id')
-        if kwargs.get('gpgkey'):
+        if kwargs.get('gpgkey') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: string length > 0

File: pymisp/api.py [hunk 9]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['gpgkey'] = kwargs.get('gpgkey')
-        if kwargs.get('certif_public'):
+        if kwargs.get('certif_public') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: same as gpg

File: pymisp/api.py [hunk 10]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
             user['certif_public'] = kwargs.get('certif_public')
-        if kwargs.get('autoalert'):
+        if kwargs.get('autoalert') is not None:
             user['autoalert'] = kwargs.get('autoalert')
-        if kwargs.get('contactalert'):
+        if kwargs.get('contactalert') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: string > 0

File: pymisp/api.py [hunk 11]
```diff
@@ -1127,41 +1127,41 @@ def get_sharing_groups(self):
...
+        if kwargs.get('change_pw') is not None:
             user['change_pw'] = kwargs.get('change_pw')
-        if kwargs.get('termsaccepted'):
+        if kwargs.get('termsaccepted') is not None:
             user['termsaccepted'] = kwargs.get('termsaccepted')
-        if kwargs.get('newsread'):
+        if kwargs.get('newsread') is not None:
             user['newsread'] = kwargs.get('newsread')
-        if kwargs.get('authkey'):
+        if kwargs.get('authkey') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: string length > 0

File: pymisp/api.py [hunk 12]
```diff
@@ -1223,23 +1223,23 @@ def delete_user(self, user_id):
 
     def _set_organisation_parameters(self, **kwargs):
         organisation = {}
-        if kwargs.get('name'):
+        if kwargs.get('name') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: str > 0

File: pymisp/api.py [hunk 13]
```diff
@@ -1223,23 +1223,23 @@ def delete_user(self, user_id):
...
             organisation['name'] = kwargs.get('name')
-        if kwargs.get('anonymise'):
+        if kwargs.get('anonymise') is not None:
             organisation['anonymise'] = kwargs.get('anonymise')
-        if kwargs.get('description'):
+        if kwargs.get('description') is not None:
```
Comment (Rafiot) [inline pymisp/api.py]: str > 0

**Comment by Rafiot** [comment]
Note that I didn't comment all the changes, but I think you meant to allow `False` for boolean entries, which is right. I don't think it is wise to allow empty string, or 0, for most of the fields.

**Comment by FloatingGhost** [comment]
You can probably rely on the falsey nature of "" and 0 to shorten everything a bit :)

**Comment by restanrm** [comment]
I've made some modifications, but I'm not sure about what your asking. Is it to write some validation functions to type check and validate user input ? It's partially done in MISP and quite franckly hard to validate. 

I'd like to write some test cases for this part of PyMISP but don't know how to play the tests locally.

File: pymisp/api.py [hunk 14]
```diff
@@ -1127,37 +1127,37 @@ def get_sharing_groups(self):
...
-        if kwargs.get('org_id'):
+        if kwargs.get('org_id') > 0:
             user['org_id'] = kwargs.get('org_id')
-        if kwargs.get('role_id'):
+        if kwargs.get('role_id') > 0:
             user['role_id'] = kwargs.get('role_id')
         if kwargs.get('password'):
             user['password'] = kwargs.get('password')
-        if kwargs.get('external_auth_required'):
+        if kwargs.get('external_auth_required') is not None:
```
Comment (FloatingGhost) [inline pymisp/api.py]: There is literally no point to these `is not None` checks.

`.get()` will return None if the key isn't found, and the value if it is.

additionally, "" is falsey. 

Hence, we can do
```python
if kwargs.get('external_auth_required'):
    user['external_auth_required'] = kwargs.get('external_auth_required')
￼```

Comment (Rafiot) [inline pymisp/api.py]: Yes, the main (only?) reason to check `None` is when the value can be `False`. 
Example:

``` python
if kwargs.get('termsaccepted') is not None: 
    user['termsaccepted'] = kwargs.get('termsaccepted')
```

With `kwargs['termsaccepted'] == False` and we want to pass that `False` to the sever.

**Comment by restanrm** [comment]
I agree with the fact that the only reason to do this is to allow `False` value to be propagated to MISP. Is there a list of field type that can be used to create a good mapping of these cases ?

Database schema are of no use, for example `change_pw` is a `tinyint(4)`
Model validation doesn't check all fields so it's not a complete informations that we can use.

I may have missed one can you help me ?

**Comment by restanrm** [comment]
I've updated commit with only boolean type and removed field that were no longer used.

**Comment by Rafiot** [comment]
I did a small patch to allow what you wanted to do, I think. 
Please let me know if it works the way you expect it to.

**Comment by restanrm** [comment]
It's ok with this, but you reintroduced fields 
* external_auth_required
* external_auth_key
* enable_password 
* certif_public

 that didn't exist in MISP database schema and produce `None` result when sent to MISP instance. Do you prefer another pull request for this modification ?

**Comment by Rafiot** [comment]
We fixed an other problem with the add_user method (on MISP HEAD). 

Which version of MISP are you using? 

The extra fields should not be a problem (they are ignored by the server): 
`external_auth_required` and `external_auth_key` are created if you enable CustomAuth, `enable_password` is only used by the UI and `certif_public` is created if you enable SMIME

**Comment by restanrm** [comment]
Ok no problem about those fields I didn't know and informations wasn't in database schema. I'm currently running on 2.4.70, I've not migrated to 2.4.71 yet.

**Comment by Rafiot** [comment]
We sorted it out today ;)

Just to make sure: the exception you are experiencing is fixed in the current HEAD (not 2.4.71) so you will have two options: 
1. pulling the current HEAD
2. waiting until the release of 2.4.72

**Pull request closed without merging**